File: blue-score-layers-audio-core/src/main/java/blue/score/layers/audio/core/AudioClip.java
/*
 * blue - object composition environment for csound
 * Copyright (C) 2013
 * Steven Yi <stevenyi@gmail.com>
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 */
package blue.score.layers.audio.core;

import blue.score.ScoreObject;
import blue.score.ScoreObjectEvent;
import blue.score.ScoreObjectListener;
import blue.time.TimeBase;
import blue.time.TimeContext;
import blue.time.TimeDuration;
import blue.time.TimePosition;
import blue.time.TimeUnitMath;
import blue.time.TimeUtilities;
import blue.utility.XMLUtilities;
import electric.xml.Element;
import electric.xml.Elements;
import java.awt.Color;
import java.beans.PropertyChangeListener;
import java.beans.PropertyChangeSupport;
import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import javax.sound.sampled.AudioFileFormat;
import javax.sound.sampled.AudioFormat;
import javax.sound.sampled.AudioSystem;
import javax.sound.sampled.UnsupportedAudioFileException;
import org.openide.util.Exceptions;

/**
 *
 * @author stevenyi
 */
public final class AudioClip implements ScoreObject, Comparable<AudioClip> {

    public static final String NAME = "name";
    public static final String START_TIME = "startTime";
    public static final String DURATION = "duration";
    public static final String COLOR = "color";
    public static final String FILE_START_TIME = "fileStartTime";
    public static final String FADE_IN = "fadeIn";
    public static final String FADE_IN_TYPE = "fadeInType";
    public static final String FADE_OUT = "fadeOut";
    public static final String FADE_OUT_TYPE = "fadeOutType";
    public static final String LOOPING = "looping";
    public static final String AUDIO_FILE = "audioFile";

    private final PropertyChangeSupport pcs = new PropertyChangeSupport(this);

    private String name = "";
    private TimePosition startTimePosition = TimePosition.beats(0.0);
    private TimeDuration durationUnit = TimeDuration.beats(0.0);
    private Color color = Color.DARK_GRAY;

    private File audioFile = null;
    int numChannels = 0;
    double audioDuration = 0.0f;

    private double fileStartTime = 0.0;
    private double fadeIn = 0.0;
    private FadeType fadeInType = FadeType.LINEAR;
    private double fadeOut = 0.0;
    private FadeType fadeOutType = FadeType.LINEAR;
    private boolean looping = true;

    transient List<ScoreObjectListener> scoreObjListeners = null;
    transient int cloneSourceHashCode = 0;

    public AudioClip() {
    }

    public AudioClip(AudioClip ac) {
        this.name = ac.name;
        this.startTimePosition = ac.startTimePosition;
        this.durationUnit = ac.durationUnit;
        this.color = ac.color;
        this.audioFile = ac.audioFile;
        this.numChannels = ac.numChannels;
        this.audioDuration = ac.audioDuration;
        this.fileStartTime = ac.fileStartTime;
        this.fadeIn = ac.fadeIn;
        this.fadeInType = ac.fadeInType;
        this.fadeOut = ac.fadeOut;
        this.fadeOutType = ac.fadeOutType;
        this.looping = ac.looping;
        this.cloneSourceHashCode = ac.hashCode();
    }

    protected void readAudioFileProperties() {
        AudioFileFormat aFormat;
        try {
            aFormat = AudioSystem.getAudioFileFormat(audioFile);
            AudioFormat format = aFormat.getFormat();

            numChannels = format.getChannels();
            audioDuration = aFormat.getByteLength()
                    / (format.getSampleRate() * (format.getSampleSizeInBits() / 8) * format
                    .getChannels());

        } catch (UnsupportedAudioFileException | IOException ex) {
            Exceptions.printStackTrace(ex);
        }
    }

    // PropertyChangeListener support

    public void addPropertyChangeListener(PropertyChangeListener listener) {
        pcs.addPropertyChangeListener(listener);
    }

    public void removePropertyChangeListener(PropertyChangeListener listener) {
        pcs.removePropertyChangeListener(listener);
    }

    // Name

    @Override
    public void setName(String value) {
        String old = this.name;
        this.name = value;
        pcs.firePropertyChange(NAME, old, value);
        fireScoreObjectEvent(new ScoreObjectEvent(this, ScoreObjectEvent.NAME));
    }

    @Override
    public String getName() {
        return name;
    }

    // Start time

    @Override
    public void setStartTime(TimePosition startTime) {
        if (startTime == null) {
            throw new IllegalArgumentException("Start time cannot be null");
        }
        TimePosition old = this.startTimePosition;
        this.startTimePosition = startTime;
        pcs.firePropertyChange(START_TIME, old, startTime);
        fireScoreObjectEvent(new ScoreObjectEvent(this, ScoreObjectEvent.START_TIME));
    }

    @Override
    public TimePosition getStartTime() {
        return startTimePosition;
    }

    // Subjective duration

    @Override
    public void setSubjectiveDuration(TimeDuration duration) {
        if (duration == null) {
            throw new IllegalArgumentException("Duration cannot be null");
        }
        TimeDuration old = this.durationUnit;
        this.durationUnit = duration;
        pcs.firePropertyChange(DURATION, old, duration);
        fireScoreObjectEvent(new ScoreObjectEvent(this, ScoreObjectEvent.DURATION));
    }

    @Override
    public TimeDuration getSubjectiveDuration() {
        return durationUnit;
    }

    // File start time

    public void setFileStartTime(double value) {
        double old = this.fileStartTime;
        this.fileStartTime = value;
        pcs.firePropertyChange(FILE_START_TIME, old, value);
    }

    public double getFileStartTime() {
        return fileStartTime;
    }

    // Fade in

    public void setFadeIn(double value) {
        double old = this.fadeIn;
        this.fadeIn = value;
        pcs.firePropertyChange(FADE_IN, old, value);
    }

    public double getFadeIn() {
        return fadeIn;
    }

    // Fade out

    public void setFadeOut(double value) {
        double old = this.fadeOut;
        this.fadeOut = value;
        pcs.firePropertyChange(FADE_OUT, old, value);
    }

    public double getFadeOut() {
        return fadeOut;
    }

    // Looping

    public void setLooping(TimeContext context, boolean looping) {
        boolean old = this.looping;
        this.looping = looping;

        if (!looping) {
            var durLimit = getAudioDuration() - getFileStartTime();
            double durBeats = durationUnit.toBeats(context);
            if (durBeats > durLimit) {
                setSubjectiveDuration(TimeUnitMath.beatsToDuration(
                        durLimit, durationUnit.getTimeBase(), context));
            }
        }

        pcs.firePropertyChange(LOOPING, old, looping);
    }

    // Simple setter for copy constructor and deserialization - no validation
    private void setLooping(boolean looping) {
        this.looping = looping;
    }

    public boolean isLooping() {
        return looping;
    }

    // Audio file

    public File getAudioFile() {
        return audioFile;
    }

    public void setAudioFile(File audioFile) {
        File old = this.audioFile;
        this.audioFile = audioFile;
        readAudioFileProperties();
        setSubjectiveDuration(TimeDuration.fromSeconds(audioDuration));
        pcs.firePropertyChange(AUDIO_FILE, old, audioFile);
    }

    public double getAudioDuration() {
        return audioDuration;
    }

    public void setAudioDuration(double originalDuration) {
        audioDuration = originalDuration;
    }

    // Color

    @Override
    public Color getBackgroundColor() {
        return this.color;
    }

    @Override
    public void setBackgroundColor(Color color) {
        Color old = this.color;
        this.color = color;
        pcs.firePropertyChange(COLOR, old, color);
        fireScoreObjectEvent(new ScoreObjectEvent(this, ScoreObjectEvent.COLOR));
    }

    public int getNumChannels() {
        return numChannels;
    }

    // Fade types

    public FadeType getFadeInType() {
        return this.fadeInType;
    }

    public void setFadeInType(FadeType fadeType) {
        FadeType old = this.fadeInType;
        this.fadeInType = fadeType;
        pcs.firePropertyChange(FADE_IN_TYPE, old, fadeType);
    }

    public FadeType getFadeOutType() {
        return this.fadeOutType;
    }

    public void setFadeOutType(FadeType fadeType) {
        FadeType old = this.fadeOutType;
        this.fadeOutType = fadeType;
        pcs.firePropertyChange(FADE_OUT_TYPE, old, fadeType);
    }

    // Resize

    @Override
    public double[] getResizeRightLimits(TimeContext context) {
        double durBeats = durationUnit.toBeats(context);
        return isLooping()
                ? new double[]{-durBeats, Double.MAX_VALUE}
                : new double[]{-durBeats, (getAudioDuration() - (durBeats + getFileStartTime()))};
    }

    @Override
    public double[] getResizeLeftLimits(TimeContext context) {
        double startBeats = startTimePosition.toBeats(context);
        double durBeats = durationUnit.toBeats(context);
        var leftLimit = isLooping()
                ? -startBeats
                : Math.max(-startBeats, -getFileStartTime());

        return new double[]{leftLimit, durBeats};
    }

    @Override
    public void resizeLeft(TimeContext context, double newStartTime) {
        double currentStart = startTimePosition.toBeats(context);
        double currentDuration = durationUnit.toBeats(context);
        double diff = currentStart - newStartTime;
        double fileStart = getFileStartTime() - diff;
        double audioDur = getAudioDuration();

        if (audioDur > 0) {
            while (fileStart < 0) {
                fileStart += audioDur;
            }
            while (fileStart > audioDur) {
                fileStart -= audioDur;
            }
        }

        setStartTime(TimeUtilities.beatsToTimePosition(
                newStartTime, startTimePosition.getTimeBase(), context));
        setFileStartTime(fileStart);
        setSubjectiveDuration(TimeUnitMath.beatsToDuration(
                currentDuration + diff, durationUnit.getTimeBase(), context));
    }

    @Override
    public void resizeRight(TimeContext context, double newEndTime) {
        double currentStart = startTimePosition.toBeats(context);
        setSubjectiveDuration(TimeUnitMath.beatsToDuration(
                newEndTime - currentStart, durationUnit.getTimeBase(), context));
    }

    @Override
    public int compareTo(AudioClip o) {
        // Compare using beat values with a default TimeContext
        var context = new TimeContext();
        int cmp = Double.compare(
                this.startTimePosition.toBeats(context),
                o.startTimePosition.toBeats(context));
        if (cmp != 0) {
            return cmp;
        }
        return Double.compare(
                this.durationUnit.toBeats(context),
                o.durationUnit.toBeats(context));
    }

    // XML Methods

    public Element saveAsXML() {
        Element root = new Element("audioClip");

        root.addElement("name").setText(getName());
        root.addElement("audioFile").setText(getAudioFile().getAbsolutePath());
        root.addElement(XMLUtilities.writeInt("numChannels", getNumChannels()));
        root.addElement(XMLUtilities.writeDouble("audioDuration",
                getAudioDuration()));
        root.addElement(XMLUtilities.writeDouble("fileStart", getFileStartTime()));
        root.addElement(getStartTime().saveAsXML().setName("startTime"));
        root.addElement(getSubjectiveDuration().saveAsXML().setName("subjectiveDuration"));
        root.addElement(XMLUtilities.writeDouble("fadeIn", getFadeIn()));
        root.addElement("fadeInType").setText(getFadeInType().toString());
        root.addElement(XMLUtilities.writeDouble("fadeOut", getFadeOut()));
        root.addElement("fadeOutType").setText(getFadeOutType().toString());
        root.addElement(XMLUtilities.writeBoolean("looping", isLooping()));

        String colorStr = Integer.toString(getBackgroundColor().getRGB());
        root.addElement("backgroundColor").setText(colorStr);

        return root;
    }

    public static AudioClip loadFromXML(Element data) {
        AudioClip clip = new AudioClip();

        Elements nodes = data.getElements();

        while (nodes.hasMoreElements()) {
            final Element node = nodes.next();
            final String nodeText = node.getTextString();

            switch (node.getName()) {
                case "name":
                    clip.setName(nodeText);
                    break;
                case "audioFile":
                    clip.setAudioFile(new File(nodeText));
                    break;
                case "numChannels":
                    clip.numChannels = XMLUtilities.readInt(node);
                    break;
                case "audioDuration":
                    clip.setAudioDuration(XMLUtilities.readDouble(node));
                    break;
                case "fileStart":
                    clip.setFileStartTime(XMLUtilities.readDouble(node));
                    break;
                case "startTime":
                    if (node.getAttributeValue("type") != null) {
                        try {
                            clip.setStartTime(TimePosition.loadFromXML(node));
                        } catch (Exception ex) {
                            Exceptions.printStackTrace(ex);
                        }
                    } else {
                        clip.setStartTime(TimePosition.beats(
                                Double.parseDouble(node.getTextString())));
                    }
                    break;
                case "start":
                    // Legacy format: plain double (beats)
                    clip.setStartTime(TimePosition.beats(XMLUtilities.readDouble(node)));
                    break;
                case "subjectiveDuration":
                    if (node.getAttributeValue("type") != null) {
                        try {
                            clip.setSubjectiveDuration(TimeDuration.loadFromXML(node));
                        } catch (Exception ex) {
                            Exceptions.printStackTrace(ex);
                        }
                    } else {
                        clip.setSubjectiveDuration(TimeDuration.beats(
                                Double.parseDouble(node.getTextString())));
                    }
                    break;
                case "duration":
                    // Legacy format: plain double (beats)
                    clip.setSubjectiveDuration(TimeDuration.beats(XMLUtilities.readDouble(node)));
                    break;
                case "backgroundColor":
                    String colorStr = data.getTextString("backgroundColor");
                    clip.setBackgroundColor(
                            new Color(Integer.parseInt(colorStr)));
                    break;
                case "fadeIn":
                    clip.setFadeIn(XMLUtilities.readDouble(node));
                    break;
                case "fadeInType":
                    clip.setFadeInType(FadeType.fromString(nodeText));
                    break;
                case "fadeOut":
                    clip.setFadeOut(XMLUtilities.readDouble(node));
                    break;
                case "fadeOutType":
                    clip.setFadeOutType(FadeType.fromString(nodeText));
                    break;
                case "looping":
                    clip.setLooping(XMLUtilities.readBoolean(node));
                    break;
            }
        }

        return clip;
    }

    // ScoreObject listener support

    @Override
    public void addScoreObjectListener(ScoreObjectListener listener) {
        if (scoreObjListeners == null) {
            scoreObjListeners = new ArrayList<>();
        }
        scoreObjListeners.add(listener);
    }

    @Override
    public void removeScoreObjectListener(ScoreObjectListener listener) {
        if (scoreObjListeners != null) {
            scoreObjListeners.remove(listener);
        }
    }

    protected void fireScoreObjectEvent(ScoreObjectEvent event) {
        if (scoreObjListeners != null) {
            for (ScoreObjectListener listener : scoreObjListeners) {
                listener.scoreObjectChanged(event);
            }
        }
    }

    @Override
    public AudioClip deepCopy() {
        return new AudioClip(this);
    }

    @Override
    public int getCloneSourceHashCode() {
        return cloneSourceHashCode;
    }

}

File: packages/blue-data/src/score/audio/audio-clip.ts
/**
 * AudioClip — a file-based audio clip in a Track.
 * Mirrors the Java AudioClip class.
 *
 * AudioClip represents a reference to an audio file with temporal positioning,
 * fade in/out, and looping settings. During CSD generation, it produces a
 * diskin2-based score event.
 */
import { ScoreObject } from '../../score/score-object';
import { TimePosition } from '../../time/time-position';
import { TimeDuration } from '../../time/time-duration';
import { TimeContext } from '../../time/time-context';
import { beatsToDuration } from '../../time/time-unit-math';
import { beatsToTimePosition } from '../../time/time-utilities';
import { FadeType, fadeTypeFromString, fadeTypeToCsound } from './fade-type';
import { Element } from '../../serialization/xml-reader';
import { ObjRefSaveMap } from '../../serialization/obj-ref-map';
import { readInt, readDouble, writeInt, writeDouble, writeBoolean } from '../../utilities/xml';

export class AudioClip implements ScoreObject {
  static readonly NAME = 'name';
  static readonly START_TIME = 'startTime';
  static readonly DURATION = 'duration';
  static readonly COLOR = 'color';
  static readonly FILE_START_TIME = 'fileStartTime';
  static readonly FADE_IN = 'fadeIn';
  static readonly FADE_IN_TYPE = 'fadeInType';
  static readonly FADE_OUT = 'fadeOut';
  static readonly FADE_OUT_TYPE = 'fadeOutType';
  static readonly LOOPING = 'looping';
  static readonly AUDIO_FILE = 'audioFile';

  private _name = '';
  private _startTimePosition = TimePosition.beats(0);
  private _durationUnit = TimeDuration.beats(0);
  private _color = 0x404040; // dark gray

  private _audioFile = '';
  private _numChannels = 0;
  private _audioDuration = 0;
  private _fileStartTime = 0;
  private _fadeIn = 0;
  private _fadeInType = FadeType.LINEAR;
  private _fadeOut = 0;
  private _fadeOutType = FadeType.LINEAR;
  private _looping = true;

  private _cloneSourceHashCode = 0;

  constructor() {}

  /** Copy constructor. */
  static copyFrom(src: AudioClip): AudioClip {
    const clip = new AudioClip();
    clip._name = src._name;
    clip._startTimePosition = src._startTimePosition;
    clip._durationUnit = src._durationUnit;
    clip._color = src._color;
    clip._audioFile = src._audioFile;
    clip._numChannels = src._numChannels;
    clip._audioDuration = src._audioDuration;
    clip._fileStartTime = src._fileStartTime;
    clip._fadeIn = src._fadeIn;
    clip._fadeInType = src._fadeInType;
    clip._fadeOut = src._fadeOut;
    clip._fadeOutType = src._fadeOutType;
    clip._looping = src._looping;
    clip._cloneSourceHashCode = src.hashCode();
    return clip;
  }

  // ─── ScoreObject ───

  getName(): string {
    return this._name;
  }
  setName(value: string): void {
    this._name = value;
  }

  getStartTime(): TimePosition {
    return this._startTimePosition;
  }
  setStartTime(value: TimePosition): void {
    this._startTimePosition = value;
  }

  getSubjectiveDuration(): TimeDuration {
    return this._durationUnit;
  }
  setSubjectiveDuration(value: TimeDuration): void {
    this._durationUnit = value;
  }

  getBackgroundColor(): number {
    return this._color;
  }
  setBackgroundColor(color: number): void {
    this._color = color;
  }

  getResizeLeftLimits(context: TimeContext): number[] {
    const startBeats = this._startTimePosition.toBeats(context);
    const durBeats = this._durationUnit.toBeats(context);
    const leftLimit = this._looping ? -startBeats : Math.max(-startBeats, -this._fileStartTime);
    return [leftLimit, durBeats];
  }

  getResizeRightLimits(context: TimeContext): number[] {
    const durBeats = this._durationUnit.toBeats(context);
    return this._looping
      ? [-durBeats, Infinity]
      : [-durBeats, this._audioDuration - (durBeats + this._fileStartTime)];
  }

  resizeLeft(context: TimeContext, newStartTime: number): void {
    const currentStart = this._startTimePosition.toBeats(context);
    const currentDuration = this._durationUnit.toBeats(context);
    const diff = currentStart - newStartTime;
    let fileStart = this._fileStartTime - diff;
    const audioDur = this._audioDuration;

    if (audioDur > 0) {
      while (fileStart < 0) fileStart += audioDur;
      while (fileStart > audioDur) fileStart -= audioDur;
    }

    this._startTimePosition = beatsToTimePosition(
      newStartTime,
      this._startTimePosition.getTimeBase(),
      context,
    );
    this._fileStartTime = fileStart;
    this._durationUnit = TimeDuration.beats(currentDuration + diff);
  }

  resizeRight(context: TimeContext, newEndTime: number): void {
    const currentStart = this._startTimePosition.toBeats(context);
    this._durationUnit = TimeDuration.beats(newEndTime - currentStart);
  }

  getCloneSourceHashCode(): number {
    return this._cloneSourceHashCode;
  }

  // ─── AudioClip-specific ───

  getAudioFile(): string {
    return this._audioFile;
  }
  setAudioFile(path: string): void {
    this._audioFile = path;
    // In Java, this reads the file to get numChannels/audioDuration.
    // In TS, caller must set these manually or we leave them at 0.
  }

  getNumChannels(): number {
    return this._numChannels;
  }
  setNumChannels(n: number): void {
    this._numChannels = n;
  }

  getAudioDuration(): number {
    return this._audioDuration;
  }
  setAudioDuration(d: number): void {
    this._audioDuration = d;
  }

  getFileStartTime(): number {
    return this._fileStartTime;
  }
  setFileStartTime(t: number): void {
    this._fileStartTime = t;
  }

  getFadeIn(): number {
    return this._fadeIn;
  }
  setFadeIn(t: number): void {
    this._fadeIn = t;
  }

  getFadeInType(): FadeType {
    return this._fadeInType;
  }
  setFadeInType(ft: FadeType): void {
    this._fadeInType = ft;
  }

  getFadeOut(): number {
    return this._fadeOut;
  }
  setFadeOut(t: number): void {
    this._fadeOut = t;
  }

  getFadeOutType(): FadeType {
    return this._fadeOutType;
  }
  setFadeOutType(ft: FadeType): void {
    this._fadeOutType = ft;
  }

  isLooping(): boolean {
    return this._looping;
  }
  setLooping(context: TimeContext | null, looping: boolean): void {
    this._looping = looping;
    if (!looping && context && this._audioDuration > 0) {
      const durLimitBeats = context.secondsToBeats(
        Math.max(0, this._audioDuration - this._fileStartTime),
      );
      const durBeats = this._durationUnit.toBeats(context);
      if (durBeats > durLimitBeats) {
        this._durationUnit = beatsToDuration(
          durLimitBeats,
          this._durationUnit.getTimeBase(),
          context,
        );
      }
    }
  }

  // ─── XML Serialization ───

  saveAsXML(_objRefMap?: ObjRefSaveMap): Element {
    const root = new Element('audioClip');

    root.addElement('name').setText(this._name);
    root.addElement('audioFile').setText(this._audioFile);
    root.addElement(writeInt('numChannels', this._numChannels));
    root.addElement(writeDouble('audioDuration', this._audioDuration));
    root.addElement(writeDouble('fileStart', this._fileStartTime));
    root.addElement(this._startTimePosition.saveAsXML().setName('startTime'));
    root.addElement(this._durationUnit.saveAsXML().setName('subjectiveDuration'));
    root.addElement(writeDouble('fadeIn', this._fadeIn));
    root.addElement('fadeInType').setText(this._fadeInType);
    root.addElement(writeDouble('fadeOut', this._fadeOut));
    root.addElement('fadeOutType').setText(this._fadeOutType);
    root.addElement(writeBoolean('looping', this._looping));
    root.addElement('backgroundColor').setText(this._color.toString());

    return root;
  }

  static loadFromXML(data: Element): AudioClip {
    const clip = new AudioClip();
    const nodes = data.getElements();

    while (nodes.hasMoreElements()) {
      const node = nodes.next();
      const text = node.getTextString();

      switch (node.getName()) {
        case 'name':
          clip._name = text;
          break;
        case 'audioFile':
          clip._audioFile = text;
          break;
        case 'numChannels':
          clip._numChannels = readInt(node);
          break;
        case 'audioDuration':
          clip._audioDuration = readDouble(node);
          break;
        case 'fileStart':
          clip._fileStartTime = readDouble(node);
          break;
        case 'startTime':
          if (node.getAttributeValue('type') !== null) {
            clip._startTimePosition = TimePosition.loadFromXML(node);
          } else {
            // Legacy: plain double (beats)
            clip._startTimePosition = TimePosition.beats(readDouble(node));
          }
          break;
        case 'start':
          // Legacy format
          clip._startTimePosition = TimePosition.beats(readDouble(node));
          break;
        case 'subjectiveDuration':
          if (node.getAttributeValue('type') !== null) {
            clip._durationUnit = TimeDuration.loadFromXML(node);
          } else {
            clip._durationUnit = TimeDuration.beats(readDouble(node));
          }
          break;
        case 'duration':
          // Legacy format
          clip._durationUnit = TimeDuration.beats(readDouble(node));
          break;
        case 'backgroundColor':
          clip._color = parseInt(data.getTextString('backgroundColor') ?? '0', 10);
          break;
        case 'fadeIn':
          clip._fadeIn = readDouble(node);
          break;
        case 'fadeInType':
          clip._fadeInType = fadeTypeFromString(text) ?? FadeType.LINEAR;
          break;
        case 'fadeOut':
          clip._fadeOut = readDouble(node);
          break;
        case 'fadeOutType':
          clip._fadeOutType = fadeTypeFromString(text) ?? FadeType.LINEAR;
          break;
        case 'looping':
          clip._looping = text.toLowerCase() === 'true';
          break;
      }
    }

    return clip;
  }

  // ─── Helpers ───

  hashCode(): number {
    return this._cloneSourceHashCode || 0;
  }
}
